---
title: "Phyloseq PreProcessing Apples"
author: "Julie Cardon"
date: "2024-03-18"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.align = "center",
                      fig.path = "./figures/02_PreProcessing/") # send any figure output to this folder 
```

# Load libraries 
```{r load-libraries}
#devtools::install_github("joey711/phyloseq")
pacman::p_load(devtools, phyloseq, tidyverse, dada2, install = FALSE)
```

# Goals

Here, we will process the data into a phyloseq object. 

- ASV table 
- Taxonomy Table 
- Track Reads (metadata)

Then, we will remove the following: 

1. Remove Chloroplasts
2. Remove Mitochondria. 
3. Removing ASVs from negative controls and also negative control samples.  
4. Evaluate accuracy by looking at the Mock Community.  
5. Remove samples without "enough" reads. 

Finally, write data file of phyloseq output, which will be stored in `data/02_PreProcessing/raw_preprocessed_phyloseq.RData`.  


# Load Data 

## ASV Table
```{r load-asv-table}
# First, load asv table
load("data/01_DADA2/ASV_counts2.RData")

# Inspect asv_tab
head(asv_tab2)[,1:5]

# Fix names 
sample_names2 <- colnames(asv_tab2)
samples_fixed2 <- apply(sapply(strsplit(basename(sample_names2), "_"), `[`,1:2), 2, paste, collapse  = "_" )
head(samples_fixed2)

# re-write the ASV count file to fix names 
colnames(asv_tab2) <- samples_fixed2
str(asv_tab2)

#save asv_tab
save(asv_tab2, file = "data/02_PreProcessing/asv_tab2.RData")
```

##  Taxonomy Table
```{r load-tax-table}
tax_df2 <- read.table("data/01_DADA2/ASV_taxonomy2.tsv", sep = "\t", skip = 1)
head(tax_df2)

# fix column names 
colnames(tax_df2) <- c("asv_names", "Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species", "ASV", "ASVseq")

head(tax_df2)

# Taxonomy Table Matrix
tax_mat2 <- 
  tax_df2 %>%
  tibble::column_to_rownames(., var = "asv_names") %>%
  as.matrix()

#Save tax_mat
save(tax_mat2, file = "data/02_PreProcessing/tax_mat2.RData")
```


##  Track Reads Data
```{r load-track-reads}
load("data/01_DADA2/track_read_counts2.RData")

# Take a look at the data
head(track_read_counts2)
dim(track_read_counts2)

# Please in the terminal copy metadata.csv 
# from: /workdir/jmc753/git_repos/AppleRhizRepository/data/metadata.csv

# Load in metadata
metadata_df2 <- read.csv("data/Jmetadata_2022.csv")
dim(metadata_df2)
colnames(metadata_df2)


# Merge metadata_df with track_reads_df 
head(metadata_df2)
head(track_read_counts2)

metadata_track_reads_df2 <- 
  metadata_df2 %>%
  left_join(., track_read_counts2, by = c("Sample" = "names"))

# Intuition check 
head(metadata_track_reads_df2)

# Update row.names to be sample names 
## Before 
row.names(metadata_track_reads_df2)
# Rewrite 
row.names(metadata_track_reads_df2) <- metadata_track_reads_df2$Sample
# Check afterwards that it worked 
row.names(metadata_track_reads_df2)
# intuition check
head(metadata_track_reads_df2)
# Save metadata_track_reads_df2
save(metadata_track_reads_df2, file = "data/02_PreProcessing/metadata_track_reads_df2.RData")
```


# Handoff to phyloseq
```{r phyloseq-handoff}
# double check it's all good 
dim(asv_tab2)
dim(tax_mat2)

# Intuition check 
stopifnot(row.names(asv_tab2) == row.names(tax_mat2))

# Construct the phyloseq object 
raw_physeq2 <- phyloseq(otu_table(asv_tab2, taxa_are_rows = TRUE),
                       sample_data(metadata_track_reads_df2),
                       tax_table(tax_mat2))
raw_physeq2

# Save this raw phyloseq object 
save(raw_physeq2, file = "data/02_PreProcessing/raw_physeq2.RData")
```

# Clean up the data

Remove: 

1. Chloroplasts
2. mitochondria  

```{r rm-mitos-chloros}
# Remind myself of tax table 
#View(tax_mat2)

# Make new physeq without chloroplasts
noChloros_physeq2 <- 
  raw_physeq2 %>% 
  # rm chloroplasts
  subset_taxa(Order != "Chloroplast" | is.na(Order))
  
# How many taxa were chloroplasts? 
num_chloro_ASVs2 <- ntaxa(raw_physeq2) - ntaxa(noChloros_physeq2)
num_chloro_ASVs2

# Intuition chek 
#noChloros_physeq2 %>%
 # tax_table2() %>%
  #data.frame() %>%
  #View()

# remove mitochondria 
noChlorosMitos_physeq2 <- 
  noChloros_physeq2 %>%
  subset_taxa(Family != "Mitochondria" | is.na(Family))

# How many mitochondrial ASVs? 
num_mito_ASVs2 <- ntaxa(noChloros_physeq2) - ntaxa(noChlorosMitos_physeq2)
num_mito_ASVs2

noChlorosMitos_physeq2

# How many total asvs were removed from chloros and mitos 
ntaxa(raw_physeq2) - ntaxa(noChlorosMitos_physeq2)
# proportion of asvs kept? 
ntaxa(noChlorosMitos_physeq2)/ntaxa(raw_physeq2)
```
## Kept 99% of the taxa

# Evaluate the Sequencing Depth 

```{r seq-depth}
# The current data object
noChlorosMitos_physeq2

# What is the library size/sequencing depth for each sample? 
seqSums_df2 <- 
  noChlorosMitos_physeq2 %>%
  otu_table() %>%
  # Sum each sample column 
  colSums() %>%
  data.frame() %>%
  rownames_to_column(var = "Sample") %>%
  left_join(., metadata_track_reads_df, by = "Sample") 

# Rename second column 
colnames(seqSums_df)[2] <- "TotalSeqs"

# check
dim(seqSums_df)
head(seqSums_df)

# Show the depth of samples 
seqSums_df %>%
  dplyr::select(Sample, TotalSeqs) %>%
  arrange(TotalSeqs) %>%
  head()

# plot it! 
seqSums_df %>%
  ggplot(aes(x=reorder(Sample, TotalSeqs), y = TotalSeqs,
             fill = Rootstock)) + 
  geom_bar(stat = "identity") 

# Density plot 
seqSums_df %>%
  ggplot(aes(TotalSeqs, fill = Rootstock)) +
  geom_density(alpha = 0.5)
```

# Remove samples with few reads 

```{r rm-samps}
# What's the min seq depth? 
min(sample_sums(noChlorosMitos_physeq))

#  Make a new object
raw_preprocessed_physeq <- 
 noChlorosMitos_physeq #%>%
  #subset_samples(., Sample != "G11_2022")

# Seperating the two years and Remove G11_2022
raw_preprocessed_physeq_2022 <- raw_preprocessed_physeq %>% subset_samples(YearSampled==2022) %>%
  subset_samples(., Sample != "G11_2022")
raw_preprocessed_physeq_2023 <- raw_preprocessed_physeq %>% subset_samples(YearSampled==2023)
#View(raw_preprocessed_physeq_2022)
#What's the new min seq depth?
min(sample_sums(raw_preprocessed_physeq))
min(sample_sums(raw_preprocessed_physeq_2022))
min(sample_sums(raw_preprocessed_physeq_2023))
```

# Save Preprocessed Phyloseq Object
```{r save-physeq}
save(raw_preprocessed_physeq,
     file = "data/02_PreProcessing/raw_preprocessed_physeq.RData")
save(raw_preprocessed_physeq_2022,
     file = "data/02_PreProcessing/raw_preprocessed_physeq_2022.RData")
save(raw_preprocessed_physeq_2023,
     file = "data/02_PreProcessing/raw_preprocessed_physeq_2023.RData")
```

```{r rarefy-phylos-object}
load(raw_preprocessed_physeq,
     file = "data/02_PreProcessing/raw_preprocessed_physeq.RData")

rare_phylos <- rarefy_even_depth(raw_preprocessed_physeq, sample.size = min(sample_sums(raw_preprocessed_physeq)), rngseed = TRUE, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)



#in inext endpoint 1000
```


# Session Information 
```{r session-info}
# Ensure reproducibility 
devtools::session_info()
```


