title: "Phylogenetic Tree Construction 2022"
author: "Julie Cardon"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---

# Goals of this file 

The goal is to create a phylogenetic tree! 

1. Load in preprocessed phyloseq object.  
2. Create ASV fasta file from the phyloseq object. 
3. Align the 16S sequences from fasta file with MAFFT. 
4. Create a tree with FastTree2.  


## Load Packages & Phyloseq Object
```{r load-environment}
# phytools, ggtree, RColorBrewer
pacman::p_load(phytools, ggtree, RColorBrewer, install = FALSE)

# Load physeq 
load("data/02_PreProcessing/raw_preprocessed_physeq_2022.RData")
raw_preprocessed_physeq_2022
```

# Create Fasta File of ASV and their Sequences

This fasta file will be used to create our alignment in MAFFT

```{r create-preprocessed-fast}
# pull out ASV seqs and ASV names 
asv_seq_df_2022 <- 
  raw_preprocessed_physeq_2022@tax_table %>%
  data.frame() %>%
  dplyr::select(ASV, ASVseq)

#View(asv_seq_df_2022)

# Add the > to make fasta header 
asv_seq_df_2022$ASV <- paste0(">",asv_seq_df_2022$ASV)
#View(asv_seq_df_2022)

# Create the fasta object 
asv_seq_fasta_2022 <- c(rbind(asv_seq_df_2022$ASV, asv_seq_df_2022$ASVseq))
head(asv_seq_fasta_2022)

# Write to a file 
write(asv_seq_fasta_2022, 
      file = "data/03_Phylogenetic_Tree/preprocessed_ASVs_2022.fasta")
```

# Align the 16S sequences from fasta file with MAFFT

```{r run-mafft, engine = 'bash', engine.opts='-l'}

# Run MAFFT on file
# I am using auto options for now (instead of specifying arguments)
# MAFFT automatically recognizes that it is a nucleotide alignment

# Set seed (not sure if relevant but not taking chances)
#RANDOM=31491

/programs/mafft/bin/mafft --auto --quiet AppleRhizRepository/data/03_Phylogenetic_Tree/preprocessed_ASVs_2022.fasta > AppleRhizRepository/data/03_Phylogenetic_Tree/aligned_ASVs_2022.fasta 

```
