---
title: "Phylogenetic Tree Inspection and Rooting"
author: "Julie Cardon"
date: "2024-04-23"
output:
  html_document: 
    code_folding: show
    theme: spacelab
    highlight: pygments
    keep_md: no
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
      toc_depth: 3
  keep_md: true  
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.align = "center",
                      # Always relevant to the document directory 
                      # send any figure output to this folder 
                      fig.path = "./figures/03_Phylogenetic_Tree/",
                      warning = FALSE) 
```


# Goals 

1. Load the fastree unrooted tree.  
2. Add tree to phyloseq object.  
3. Visualize and inspect tree with ggtree. 
4. Prune ASVs, if needed.  
5. Root our tree. 
6. Combine new tree with a phyloseq object. 
7. Save 2 phyloseq objects: 1. Unrooted tree phyloseq object, 2. Rooted tree phyloseq object. 

# Before you start

## Set my seed 
```{r set-seed}
# Any number can be chosen 
set.seed(238428)
```

## Load Packages 
```{r load-packages}
pacman::p_load(tidyverse, phyloseq, ggtree, phytools,
               install = FALSE)
```

## Load Data files 
```{r load-data}
# Preprocessed phyloseq object 
load("data/02_PreProcessing/raw_preprocessed_physeq.RData")
raw_preprocessed_physeq

# Load in the tree! 
unrooted_tree <- read.tree("data/03_Phylogenetic_Tree/ASVs_unrooted.tree")
unrooted_tree
str(unrooted_tree)
```


# Merge Phyloseq 
```{r merge-physeq}
# Intuition check 
stopifnot(ntaxa(raw_preprocessed_physeq) == ntaxa(unrooted_tree))

# Merge the tree with the phyloseq object 
unrooted_physeq <- 
  merge_phyloseq(raw_preprocessed_physeq, unrooted_tree)
unrooted_physeq
```

# Plot Tree with `ggtree`

```{r plot-tree-unrooted}
# Make a basic tree
kingdom_tree <- 
  ggtree(unrooted_physeq) + 
  # color tips by kingdom 
  geom_tippoint(mapping = aes(color = Kingdom)) + 
  scale_color_manual(values = c("goldenrod1", "cornflowerblue", "grey")) +
  # Add title 
  labs(title = "Unrooted Tree") + 
  #move the legend to the bottom 
  theme(legend.position = "bottom"); kingdom_tree

kingdom_node_tree <- 
  kingdom_tree + 
  # Add the node label 
  geom_text(aes(label=node), hjust= 1.5, vjust = -0.3, size = 2)
kingdom_node_tree
```

# Evaluate Long Branch 

This part gets a bit manual 

```{r eval-long-branch}
# View a specific clade 
# Zoom in on origin tree: Node 88977

# Zoom in on origin tree: Node 88977
viewClade(kingdom_node_tree + 
          labs(title = "Unrooted Tree: Node 88977"), 
          node = 88977)

viewClade(kingdom_node_tree + 
          labs(title = "Unrooted Tree: Node 88977") + 
          geom_text(aes(label=ASV)), 
          node = 88977)

```

4 ASVs are suspect. Let's explore these ASVs more! 


```{r check-taxonomy}
# Let's make a note of the the taxonomy and the ASV Sequence.... 
sus_ASVs <- c("ASV_39975", "ASV_58083", "ASV_35582", "ASV_59336")

unrooted_physeq %>%
  subset_taxa(., ASV %in% sus_ASVs) %>%
  tax_table() %>%
  data.frame()

# Let's also check the counts of the ASV 
unrooted_physeq %>%
  subset_taxa(., ASV %in% sus_ASVs) %>%
  otu_table() %>%
  data.frame() %>%
  colSums()

```


Let's BLAST a couple on NSCBI Nucleotide Blast 

ASV_58083 matches with: Uncultured bacterium clone IZ1RPV403DDB2F 16S ribosomal RNA gene, partial sequence with max and total score of 654, and 100% query cover, and E value of 0.0 and 94.43% Identity, and accession number KP945559.1

ASV_39975 matches with: Select seq MT626826.1	Pseudomonas sp. strain SeaQual_P_B791/7 16S ribosomal RNA gene, partial sequence	Pseudomonas sp.	total score of 785, 99% query cover, and E value of	0.0, and	100.00%	Identity, and accession number 1387	MT626826.1 


This whole clade looks suspicious because they are so far separated from the others and the contaminants need to be removed from the dataset. Let's go ahead and do that... 

# Prune the 4 suspect ASVs 
```{r prune-sus_ASVs}
# Function from Joey McMurdie: https://github.com/joey711/phyloseq/issues/652
pop_taxa = function(physeq, badTaxa){
  allTaxa <-  taxa_names(physeq)
  allTaxa <- allTaxa[!(allTaxa %in% badTaxa)]
  return(prune_taxa(allTaxa, physeq))}

# Let's use the pop_taxa function :) 
# Recreate a phyloseq object without ASV_456
unrooted_physeq_rm_sus_ASVs <- 
  unrooted_physeq %>%
  pop_taxa(., sus_ASVs)

# Intuition Check
ntaxa(unrooted_physeq) - ntaxa(unrooted_physeq_rm_sus_ASVs) 

# Visually inspect 
ggtree(unrooted_physeq_rm_sus_ASVs) + 
  geom_tippoint(mapping = aes(color = Kingdom))
```

# Midroot Tree
```{r midroot-tree}
# Is the tree rooted?
new_unrooted_tree <-phy_tree(unrooted_physeq_rm_sus_ASVs)
is.rooted(new_unrooted_tree)

# Let's midpoint root the tree
midpoint_rooted_tree <- midpoint.root(new_unrooted_tree)

# Is the new tree rooted?
is.rooted(midpoint_rooted_tree)

# Assign to a new phyloseq object: merging subsetted phyloseq with the new rooted tree

# 1. Create the phyloseq object without strange ASVs
physeq_rm_susASVs <- 
  raw_preprocessed_physeq %>%
  subset_taxa(ASV != sus_ASVs)

# Merge tree with the new physeq_rm_susASVs
midroot_physeq_rm_susASVs <- 
  merge_phyloseq(physeq_rm_susASVs, midpoint_rooted_tree)
midroot_physeq_rm_susASVs

# Quick inspection of tree 
ggtree(midroot_physeq_rm_susASVs) + 
  geom_tippoint(mapping = aes(color = Phylum))
```


# Save to a new phyloseq object
```{r save-physeq}
# Save both phyloseq objects with our tree object to one .RData file 
save(list = c("unrooted_physeq_rm_sus_ASVs", "midroot_physeq_rm_susASVs"),
     file = "data/03_Phylogenetic_Tree/phytree_preprocessed_physeq.RData")
```

# Session Information 
```{r session-info}
# Ensure reproducibility 
devtools::session_info()
```